pkg/rhcos/release/release.go

```diff
@@ -0,0 +1,180 @@
+// Package release extracts the RHCOS build version from an
+// OpenShift release image.
+package release
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"io"
+	"io/ioutil"
+
+	"github.com/containers/image/docker"
+	"github.com/containers/image/image"
+	"github.com/containers/image/pkg/blobinfocache"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+	"github.com/docker/distribution/reference"
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+// RHCOSBuild extracts the RHCOS build version from an OpenShift
+// release image.
+func RHCOSBuild(ctx context.Context, releaseImagePullSpec string, pullSecret []byte) (string, error) {
+	imageReferencesBytes, err := imageReferences(ctx, releaseImagePullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+
+	pullSpec, err := osContentPullSpec(imageReferencesBytes)
+	if err != nil {
+		return "", err
+	}
+
+	imageSource, image, err := pull(ctx, pullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+	defer imageSource.Close()
+
+	inspect, err := image.Inspect(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	version, ok := inspect.Labels["version"]
+	if !ok {
+		return "", errors.Errorf("%s has no version annotation", pullSpec)
+	}
+
+	return version, nil
+}
+
+// pull pulls an image.  If the pull secret includes credentials for
+// the pull-spec authority, those are used to authenticate requests.
+// The signature on the returned image, if any, is validated agaist
+// the signature policy from /etc/containers/policy.json.
+func pull(ctx context.Context, pullSpec string, pullSecret []byte) (types.ImageSource, types.Image, error) {
+	named, err := reference.ParseNamed(pullSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reference, err := docker.NewReference(named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sys := &types.SystemContext{}
+	err = addPullSecret(sys, pullSecret, named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	imageSource, err := reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		if err != nil {
+			imageSource.Close()
+		}
+	}()
+
+	unparsed := image.UnparsedInstance(imageSource, nil)
+
+	policy, err := signature.DefaultPolicy(sys)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
```

Comment:
> OTOH, this makes much less difference with the `containers_image_openpgp` build tag...

I think I'll just punt on this then, since it saves me from pulling the whole `SystemContext` abstraction up into `RHCOSBuild`.  But I can pull all that up if folks have strong feelings about it ;).

Reply:
As long as you are not using the GPG backend, and not trying to squeeze out every second out of the run time, this is good enough.